lib/oli/automation_setup.ex

```diff
@@ -0,0 +1,354 @@
+  end
+
+  defp create_author(true) do
+    password = random_password()
+
+    {:ok, user} =
+      Oli.Accounts.create_author(%{
+        email: generate_email("author"),
+        given_name: "Test",
+        family_name: "Author",
+        password: password,
+        password_confirmation: password,
+        system_role_id: Oli.Accounts.SystemRole.role_id().author,
+        email_confirmed_at: Timex.now()
+      })
+
+    {:ok, user, password}
+  end
+
+  defp has_project(nil) do
+    {:error, "Project not found"}
+  end
+
+  defp has_project(project) do
+    {:ok, project}
+  end
+
+  defp has_no_authors([]) do
+    {:ok}
+  end
+
+  defp has_no_authors(_) do
+    {:error, "Can only delete projects with no authors"}
+  end
+
+  defp can_delete_section(nil) do
+    {:error, "Section not found"}
+  end
+
+  defp can_delete_section(%{:title => "Automation test section"} = section) do
+    {:ok, section}
+  end
+
+  defp can_delete_section(_) do
+    {:error, "Can only delete sections with title 'Automation test section'"}
+  end
+
+  defp no_duplicates(project) do
+    child_project_exists =
+      Repo.exists?(
+        from p in Project,
+          where: p.project_id == ^project.id,
+          select: p
+      )
+
+    if not project.allow_duplication and not child_project_exists do
+      {:ok}
+    else
+      # If we have child projects, or the project allows duplicates, don't let us delete it.
+      {:error, "Project allows duplicates"}
+    end
+  end
+
+  # Tears down a test project
+  # Deletes:
+  # project
+  #   -> publications
+  #       -> publication_resources
+  #   -> project_resources
+  #       -> resources
+  #          -> revisions
+  #
+  # Only works on automation-test projects
+  def teardown_project(slug) do
+    with {:ok, project} <-
+           Repo.one(
+             from p in Project,
+               where: p.slug == ^slug,
+               preload: [
+                 :authors,
+                 :family,
+                 :resources,
+                 :activity_registrations,
+                 :part_component_registrations,
+                 :communities,
+                 :publications
+               ]
+           )
+           |> has_project(),
+         # Only tear down projects where duplicates are not allowed
+         {:ok} <- no_duplicates(project),
+         # Only tear down projects with no authors
+         {:ok} <-
+           has_no_authors(project.authors) do
+      #
+      # Delete all the publications for this project, this will also get the publication_resources
+      Enum.each(project.publications, &Oli.Publishing.delete_publication/1)
+
+      # Delete all the associations between this project and resources
+      Repo.delete_all(from pr in ProjectResource, where: pr.project_id == ^project.id, select: pr)
+
+      # Delete the resources
+      for resource <- project.resources do
+        Repo.delete_all(from r in Oli.Resources.Revision, where: r.resource_id == ^resource.id)
```

Comment:
This seems to be the most potentially dangerous part of this PR. As far as I can tell, the system wasn't really designed to ever delete revisions. With the checks in place to make sure we're only deleting data from objects created through the automation setup api, I think this is safe, but lack some understanding of the overall design to know for sure.

Reply:
While working on improving performance in torus queries I ran into some issues when querying the revisions table. It might be potentially big, and without an index on `resource_id` it could do a seq scan through the whole table.

Reply:
Without an index on resource_id, this turned out to be trickier than I hoped. I refactored the code to do everything in two delete_all calls instead of a loop like previous. This brings the number of seq-scans of the table from the number of resources to just once, but that's as good as I could get it without that index.

I populated a database with 600 imports of a project. After doing that, the tear-down takes 5-10 seconds which seems reasonable for an automated test api call.

Reply:
Looking at priv/repo/migrations/20220317164007_fix_previous_migration.exs, there's also a database trigger on the view which probably needs to be dropped/created as part of this. I think this migration would end up looking something like

```
Oli.Repo.Migrations.FixPreviousMigration.drop_trigger()
Oli.Repo.Migrations.FixPreviousMigration.drop_materialized_view()

drop the old constraint
add the new constraint with the on_delete option

Oli.Repo.Migrations.FixPreviousMigration.user = get_current_db_user()
Oli.Repo.Migrations.FixPreviousMigration.create_materialized_view(user)
Oli.Repo.Migrations.FixPreviousMigration.create_trigger(user)
Oli.Repo.Migrations.FixPreviousMigration.refresh_materialized_view()
```

Does this sound like the right path?